flax/jax_utils.py

```diff
@@ -243,3 +243,81 @@ def body_wrapper(c, xs):
   c, ys = _scan_nd(body_wrapper, init, xs, n=len(axis), unroll=unroll)
   ys = jax.tree_map(transpose_out, ys)
   return c, ys
+
+
+# Copied from https://github.com/google-research/big_vision
+def pad_shard_unpad(wrapped, static_argnums=(0,), static_argnames=(),
+                    static_return=False):
+  """Wraps a function with code that pads, shards, then un-shards, un-pads.
+
+  Args:
+    wrapped: the function to be wrapped. Signature is `params, *args, *kwargs`.
+    static_argnums: indices of arguments to `wrapped` that should _not_ be
+      padded and sharded, but instead be forwarded as-is. The default is (0,)
+      because by far the most common use-case is to pass `params` first.
+    static_argnames: names of kwargs to `wrapped` that should _not_ be padded
+      and sharded, but instead be forwarded as-is.
+    static_return: whether not to un-shard, and un-pad the return value; static
```

Comment:
Wondering if `static_return` should also accepted a tuple of ints so in case there are multiple outputs you can choose which ones to let pass untouched. Maybe this could be a future improvement.

Reply:
I considered this, but couldn't come up with a usecase. With evaluation, usually we
- either want to get the values to do something with them (e.g. custom metric computation outside JAX)
- or return metrics from an `eval_step()`

Note that for training the padding is not needed.

So I thought YAGNI and went for the simpler solution.

Reply:
Yeah, I also couldn't come up with a scenario where it was needed. Could be expanded in the future if someone finds the need.